### PR TITLE
sql/parser: implement TypeCheck, Eval as methods

### PR DESCRIPTION
**Pull request by tamird**
Work toward #2790.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/2883)

### CODE REVIEW

**Comment by tschottdorf** [comment]
what was the magic bug?

**Comment by tamird** [comment]
DReference (!!!!)

File: sql/group.go [hunk 1]
```diff
@@ -309,18 +309,28 @@ func checkAggregateExpr(expr parser.Expr) error {
 
 type aggregateValue struct {
 	datum parser.Datum
-	// Tricky: we embed a parser.FuncExpr so that aggregateValue implements
```
Comment (tamird) [inline sql/group.go]: @tschottdorf magic bug here

**Comment by petermattis** [comment]
I think this is a nice cleanup. Looking forward to seeing this done for normalization as well.

LGTM


---

Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/parser/eval.go, line 1128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EsjHKUJfrJZZcB_v0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/eval.go#L1128)):</sup>
While true, it would be easy to do: `return Tuple(t).Eval(ctx)`.


---

<sup>**[sql/parser/eval.go, line 1146 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1Esz2q737Qv1xBxfoU)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/eval.go#L1146)):</sup>
Error messages should not be capitalized and generally do not end with punctuation.


---

<sup>**[sql/parser/type_check.go, line 318 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EtZwmH0cVRfHUj4h4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L318)):</sup>
Can you expand on this? How is this a lie?


---

<sup>**[sql/parser/type_check.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1Etvs9XEWHrSJPhooa)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L416)):</sup>
See early comment: `return Tuple(expr).TypeCheck()`


---

<sup>**[sql/parser/type_check.go, line 434 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EtsjouOs8RW0Hguqm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L434)):</sup>
See earlier comment about capitalization and punctuation.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by tamird** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/parser/eval.go, line 1128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EsjHKUJfrJZZcB_v0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/eval.go#L1128)):</sup>
The compiler buys it, so done.


---

<sup>**[sql/parser/eval.go, line 1146 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1Esz2q737Qv1xBxfoU)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/eval.go#L1146)):</sup>
Yeah, this was a copy-pasta from the comments.


---

<sup>**[sql/parser/type_check.go, line 318 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EtZwmH0cVRfHUj4h4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L318)):</sup>
You're right; removed.


---

<sup>**[sql/parser/type_check.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1Etvs9XEWHrSJPhooa)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L416)):</sup>
Done.


---

<sup>**[sql/parser/type_check.go, line 434 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1EtsjouOs8RW0Hguqm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4a9810f0aaa1d9262e6c5748036aa8ba4504f50a/sql/parser/type_check.go#L434)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by petermattis** [comment]
Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[sql/parser/datum.go, line 620 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1F6Zd2PyIZdKoK0iKS)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfacf0be7700ebd90549fd2712862a39329ad02/sql/parser/datum.go#L620)):</sup>
Wonder if this should be named `VariableExpr` instead in order to give it a positive attitude.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by petermattis** [comment]
Reviewed 11 of 17 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[sql/parser/datum.go, line 620 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1F6Zd2PyIZdKoK0iKS)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfacf0be7700ebd90549fd2712862a39329ad02/sql/parser/datum.go#L620)):</sup>
Can do.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by petermattis** [comment]
LGTM


---

Review status: 12 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[sql/parser/datum.go, line 622 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1FHgC1Vmqydy5ei1pq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6788656d95486ddd52ecacf4b34413eb32486768/sql/parser/datum.go#L622)):</sup>
`s/NonConst/Variable/g`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Comment by tamird** [comment]
Review status: 12 of 18 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[sql/parser/datum.go, line 622 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883#-K1FHgC1Vmqydy5ei1pq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6788656d95486ddd52ecacf4b34413eb32486768/sql/parser/datum.go#L622)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2883)

**Pull request merged**